### ISSUE TITLE
絵文字のブラックリスト設定を追加し、リストに入っている絵文字リアクションを❤に変換できるようにしたい

### ISSUE THREAD

**Issue by EbiseLutica** [syuilo/Misskey#11772]
Unicode、カスタム絵文字問わず、人によってリアクションされると嫌な絵文字というものがあります

* ユーザー設定で、この嫌な絵文字の一覧をブラックリストとして登録できる機能
* ブラックリストにある絵文字が、自分のノートに対するリアクションとして使われた場合、リアクション絵文字を❤に変換する

の2つがあると良さそうです

（ブラックリストっていう表記、なにかとまずいかもしれない）

**Comment by tai-cha** [comment]
ブラックリスト的なものが実際にできるなら自分のTL上のリアクションからも表示しないようにできたらうれしいと私的には思います（インスタンスによってはチカチカする絵文字も入っていたりするがそれらリアクションを見ないことがむずかしいため）

**Comment by kakkokari-gtyih** [comment]
ローカルはどうにかなっても、リモートからのカスタム絵文字のブロックは際限ないのでキツそう

**Comment by kakkokari-gtyih** [comment]
> ローカルはどうにかなっても、リモートからのカスタム絵文字のブロックは際限ないのでキツそう

同一名称の絵文字は全ブロックとかにしておけばちょっとはマシになるかな

**Comment by kakkokari-gtyih** [comment]
> #12480 が実現できれば名前の異なる見たくない絵文字をミュートしたりできるのではと思ってる

https://github.com/misskey-dev/misskey/issues/12480#issuecomment-1862410762 に書いたけど、厳格な判断をしたい場合にこの手法は使い物にならなくなるかもしれない

**Comment by kakkokari-gtyih** [comment]
> ローカルはどうにかなっても、リモートからのカスタム絵文字のブロックは際限ないのでキツそう

リモートからのリアクションは全フォールバックにする設定があるのでそれとの合せ技でいいかもと思ったりした

**Comment by kakkokari-gtyih** [comment]
中途半端にバックエンドだけ実装した

https://github.com/kakkokari-gtyih/misskey/tree/feat-11772

**Comment by samunohito** [comment]
ほしいかも
着手はしばらく後になるけど、それでも良ければ僕が実装します

自分が実装するとしたら…以下のような感じで考えてます
- 禁止ワードとかと同じように、キーワードと正規表現パターンをテキストボックスに書き連ねていく
- 対象は「自分のノートにつけられたリアクション」のみ
- 対象の絵文字一覧はuser_profileに持つ
- NoteEntityServiceでのpack時に、ヒットした絵文字を❤に寄せる

**Comment by syuilo** [comment]
> NoteEntityServiceでのpack時に、ヒットした絵文字を❤に寄せる

NoteをpackするときにUserProfileのfetchが必要になるのは避けたいわね

**Comment by syuilo** [comment]
通知とかも考慮したいしクライアント側でケアするのはうーむ

**Comment by syuilo** [comment]
あと「自分だけが見たくない」というよりは「自分の投稿に対して使ってほしくない」ことの方が多そうだからサーバーサイドでやる必要がある

**Comment by samunohito** [comment]
> NoteをpackするときにUserProfileのfetchが必要になるのは避けたいわね

たしか、user_profileはキャッシュもあったはず…
あと、packMany()で呼ばれることがほとんどだろうから、many側で必要なデータをあらかじめ取り寄せておいて、packのオプションとして引き渡して…みたいなのを考えていました（NodeEntityServiceとかそのあたりが既にやってる方式）。

ともかく、DBへの負担を最小限にする実装の模索は必要です

**Comment by samunohito** [comment]
（別案として、リアクションをつけるときに❤にしてしまうというのも考えてはいるが、すでについてしまったリアクションをケア出来ないので一長一短な感じ）

**Comment by syuilo** [comment]
設定すると、過去のも含め返還されるのではなく、設定後に行われるリアクションを変換するようにすれば負荷は減りそう
(過去のものをどうにかしたい場合は https://github.com/misskey-dev/misskey/issues/14660 を使ってもらう)

**Comment by Sayamame-beans** [comment]
確かに、以降のリアクション自体を制限する11772、特定の絵文字の方を非表示にする14785(or 14660?)の組み合わせで、需要の多くはカバー出来る気もしてきますね

**Comment by Sayamame-beans** [comment]
revisit: 絵文字ミュート機能でカバー出来ている?
